### PR TITLE
Add URL safe option to Base64.

### LINKED ISSUE #30: base64 url safe encoder
Currently, the base64 encoder is not url safe. It would be nice to have an option for that. According to the [RFC 4648](https://tools.ietf.org/html/rfc4648#page-7), you just have to replace the "+" and "/" [here](https://github.com/korlibs/krypto/blob/master/krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt#L4) with "-" and "_".

### PR DESCRIPTION
**Pull request by IgnacioCarrionN**
Fixes #30 
- Add methods to encode/decode Base64Url.
- Test Base64 extension functions.

### CODE REVIEW

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 1]
```diff
@@ -1,42 +1,49 @@
...
+                TABLE_SAFE
+            } else {
+                TABLE
+            }
+            for (n in 0..255) this[n] = -1
+            for (n in table.indices) {
+                this[table[n].toInt()] = n
+            }
+        }
+
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:32]: This data is immutable after setting it, so can be in an object. Better to extract this as part of the object, and do the if with the two precomputed DECODE tables.

File: krypto/src/commonTest/kotlin/com/soywiz/krypto/encoding/Base64Test.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package com.soywiz.krypto.encoding
+
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class Base64Test {
```
Comment (soywiz) [inline krypto/src/commonTest/kotlin/com/soywiz/krypto/encoding/Base64Test.kt:6]: Cool! Thanks!

**Review by soywiz — CHANGES_REQUESTED** [review]
Can you extract it to be part of the object so the table is allocated and computed on each decode call?

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 1 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonTest/kotlin/com/soywiz/krypto/encoding/Base64Test.kt (hunk 2 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonTest/kotlin/com/soywiz/krypto/encoding/Base64Test.kt:6]: 👍

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 3]
```diff
@@ -86,6 +98,10 @@ object Base64 {
 }
 
 fun String.fromBase64IgnoreSpaces(): ByteArray = Base64.decode(this.replace(" ", "").replace("\n", "").replace("\r", ""))
+fun String.fromBase64UrlIgnoreSpaces(): ByteArray = Base64.decode(this.replace(" ", "").replace("\n", "").replace("\r", ""), true)
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:110]: To be consistent: `fromBase64UrlIgnoreSpaces` -> `fromBase64UrlSafeIgnoreSpaces`

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 4]
```diff
@@ -49,31 +56,36 @@ object Base64 {
     }
 
     @Suppress("UNUSED_CHANGED_VALUE")
-    fun encode(src: ByteArray): String {
+    fun encode(src: ByteArray, isUrlSafe: Boolean = false): String {
+        val table = if (isUrlSafe) {
+            TABLE_SAFE
+        } else {
+            TABLE
+        }
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:73]: Looking at this. What would you think about renaming TABLE and TABLE_SAFE to: `ENCODE_TABLE` and `ENCODE_TABLE_URLSAFE` to make it clearer?

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 5]
```diff
@@ -1,42 +1,58 @@
...
+        for (n in TABLE_SAFE.indices) {
+            this[TABLE_SAFE[n].toInt()] = n
+        }
+    }
+
     operator fun invoke(v: String) = decodeIgnoringSpaces(v)
     operator fun invoke(v: ByteArray) = encode(v)
 
-    fun decode(str: String): ByteArray {
+    fun decode(str: String, isUrlSafe: Boolean = false): ByteArray {
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:24]: I'm used to see the `is` as "asking", usually for a boolean getter, and here I would use urlSafe or at least doUrlSafe. Do you have a different criteria there?
Alternatively, why don't just create a base private decode method receiving the right table as parameter, and then two methods one for normal and another for url-safe? That way no ifs would be required for selecting the table

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 6]
```diff
@@ -1,42 +1,58 @@
 package com.soywiz.krypto.encoding
 
 object Base64 {
-    private val TABLE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
-    private val DECODE = IntArray(0x100).apply {
+    private const val TABLE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
+    private const val TABLE_SAFE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_="
+
+    private val decoder = IntArray(0x100).apply {
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:7]: Even if not const, this is still immutable and I see it as a constant table. I would use UPPERCASE here, or at least would be consisent for the TABLE names. Though if you did it with a criteria I'm open to see your point

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 7]
```diff
@@ -1,42 +1,58 @@
...
         for (n in 0..255) this[n] = -1
-        for (n in 0 until TABLE.length) {
+        for (n in TABLE.indices) {
             this[TABLE[n].toInt()] = n
         }
     }
 
+    private val decoderUrlSafe = IntArray(0x100).apply {
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:14]: Though I saw that initially this was not done, maybe IntArray(0x100) { -1 } would remove the requirement for the first for on both cases. Not a big body block, but maybe this table could be generated in a private function receiving the table so you can deduplicate that code

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 3 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 4 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 5 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 6 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 7 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 8]
```diff
@@ -15,28 +20,39 @@ object Base64 {
...
         val src = ByteArray(str.length) { str[it].toByte() }
         val dst = ByteArray(src.size)
-        return dst.copyOf(decode(src, dst))
+        return dst.copyOf(decode(src, dst, DECODER))
+    }
+
+    fun decodeUrlSafe(str: String): ByteArray {
+        val src = ByteArray(str.length) { str[it].toByte() }
+        val dst = ByteArray(src.size)
+        return dst.copyOf(decode(src, dst, DECODER_URLSAFE))
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:29]: To deduplicate this:

```
fun decode(str: String): ByteArray = decode(str, DECODER)
fun decodeUrlSafe(str: String): ByteArray = decode(str, DECODER_URLSAFE)

private fun decode(str: String, decoder: IntArray): ByteArray {
        val src = ByteArray(str.length) { str[it].toByte() }
        val dst = ByteArray(src.size)
        return dst.copyOf(decode(src, dst, decoder))
}
```

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt [hunk 9]
```diff
@@ -15,28 +20,39 @@ object Base64 {
...
     }
 
     fun decodeIgnoringSpaces(str: String): ByteArray {
         return decode(str.replace(" ", "").replace("\n", "").replace("\r", ""))
     }
 
-    fun decode(src: ByteArray, dst: ByteArray): Int {
+    fun decodeUrlSafeIgnoringSpaces(str: String): ByteArray {
+        return decodeUrlSafe(str.replace(" ", "").replace("\n", "").replace("\r", ""))
+    }
```
Comment (soywiz) [inline krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt:38]: To deduplicate this:

```
fun decodeIgnoringSpaces(str: String): ByteArray = decode(str.removeSpaces())
fun decodeUrlSafeIgnoringSpaces(str: String): ByteArray = decodeUrlSafe(str.removeSpaces())
private fun String.removeSpaces() = this.replace(" ", "").replace("\n", "").replace("\r", "")
```

**Review by soywiz — COMMENTED** [review]
Looks pretty good already to me. But can you deduplicate the new decode method variants?

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 8 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

File: krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt (hunk 9 above)
Comment (IgnacioCarrionN) [reply krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt]: Done

**Comment by soywiz** [comment]
Thank you very much Ignacio!

**Pull request merged**